### PR TITLE
Track number of optimised regexp label matchers

### LINKED ISSUE #4638: Track number of optimised regexp label matchers
In `mimir-prometheus` we employ several tricks to optimise the regexp label matcher, using [`FastRegexMatcher`](https://github.com/grafana/mimir-prometheus/blob/main/model/labels/regexp.go#L41).

Since in the query-frontend we parse the query, would be nice to track the total number of regex matchers found and how many of them have been optimized by `FastRegexMatcher` (we may need to expose `FastRegexMatcher.IsOptimized()`). This way we can compute the % of regexp matchers optimized and act if it goes below a critical threshold (we expect a large number of them being optimized).

### PR DESCRIPTION
**Pull request by dhanusaputra**
#### What this PR does

- add regexpMatcherCount and regexpMatcherOptimizedCount metrics
- update `longestRegexpMatcherBytes` so it can return more data

#### Which issue(s) this PR fixes or relates to

Fixes #4638 

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@
 * [ENHANCEMENT] Go: update to 1.20.3. #4773
 * [ENHANCEMENT] Store-gateway: reduce memory usage in some LabelValues calls. #4789
 * [ENHANCEMENT] Store-gateway: add a `stage` label to the metric `cortex_bucket_store_series_data_touched`. This label now applies to `data_type="chunks"` and `data_type="series"`. The `stage` label has 2 values: `processed` - the number of series that parsed - and `returned` - the number of series selected from the processed bytes to satisfy the query. #4797
+* [ENHANCEMENT] Query-frontend: track number of optimised regexp label matchers. #4813
```
Comment (56quarters) [inline CHANGELOG.md:13]:
```suggestion
* [ENHANCEMENT] Query-frontend: Add `cortex_frontend_regexp_matcher_count` and `cortex_frontend_regexp_matcher_optimized_count` metrics to track optimization of regular expression label matchers. #4813
```

**Review by 56quarters — APPROVED** [review]
Great work, thanks! Just one small suggestion.

**Review by pracucci — COMMENTED** [review]
Thanks for working on this! My only comment is that that way it's implemented means it gets tracked only for tenants with query sharding enabled. I think we should track for every query. For this reason, I think a better place where to place this logic is `pkg/frontend/querymiddleware/stats.go`. You will need to parse the query there again (in a follow up PR we can do a refactoring to parse the query once and pass around the parsed query along with the `Request`).

**Comment by dhanusaputra** [comment]
test:

image

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Comment by pracucci** [comment]
I pushed a commit https://github.com/grafana/mimir/pull/4813/commits/5b6e6f95d23ad7392ab350f729fd1040ed4b090e to ignore any parsing error in `queryStatsMiddleware`, to let this PR moving forward. I plan to refactoring the query parsing anyway, doing it only once in the query-frontend, so that will be addressed in a follow up PR.

**Comment by pracucci** [comment]
@dhanusaputra could you sign the CLA (see message above)? Otherwise we can't merge this PR.

**Pull request merged**